api/app/controllers/charges_controller.rb

```diff
@@ -41,4 +44,19 @@ def charge_params
     params.require(:line_items)
     params.permit(:merchant_id, line_items: [[:amount, :currency, :name, :quantity, :description]])
   end
+
+  def validate(line_item:)
+    [:amount, :currency, :name, :quantity].each do |attribute|
+      unless line_item.key?(attribute)
+        raise ActionController::ParameterMissing.new "param is missing or the value is empty: #{attribute}"
+      end
+    end
+
+    unless ['Gift Card', 'Donation'].include? line_item['name']
+      raise InvalidLineItem.new "line_item must be named `Gift Card` or `Donation`"
+    end
+
+    amount = line_item['amount'].to_i
```

Comment:
`to_i` won't throw if it's not an integer, so if the front end starts sending you dollar amounts with decimals, this won't catch it